File: leetcode/stack/car_fleet.py
from typing import List

class Solution:
    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:
        combined = [
            [position[i], speed[i], (target - position[i]) / speed[i]]
            for i in range(len(position))
        ]

        sorted_cars = sorted(combined, key=lambda x: x[0])

        stack = []

        for i in range(len(position) - 1, -1, -1):
            car = sorted_cars[i]

            if len(stack) == 0 or stack[-1][2] < car[2]:
                stack.append(car)

        return len(stack)

# Try it out — press Run
sol = Solution()
print(sol.carFleet(10, [1,4], [3,2]))       # 1
print(sol.carFleet(10, [4,1,0,7], [2,2,1,1]))  # 3

File: leetcode/stack/car_fleet.ts
function carFleet(target: number, position: number[], speed: number[]): number {
    let combined: [number, number, number][] = position.map((pos, index) => [pos, speed[index], (target - pos) / speed[index] ]);

    // 0 = position, 1 = speed, 2 = time
    combined.sort((a, b) => b[0] - a[0]);

    let stack: [number,number,number][] = [];

    for (let i: number = 0; i < position.length; i ++) {
        let car = combined[i];
        if (stack.length == 0 || stack[stack.length - 1][2] < car[2]) {
            stack.push(car);
        }
    }

    return stack.length;
};

// Try it out — press Run
console.log(carFleet(10, [1,4], [3,2]));          // 1
console.log(carFleet(10, [4,1,0,7], [2,2,1,1]));  // 3

export {};
